File: XTEA/C++/XTEA with OS.cpp
#include <Crypto.h>
#include <Arduino.h>
#include "hardware/clocks.h"
#include "pico/bootrom.h"
#include "pico.h"
#include <string.h>
#include <FreeRTOS.h>
#include <semphr.h>
#define XTEA_ROUNDS 48
#define XTEA_MAC_ROUNDS 24
#include "XTEA-Cipher.h"

#define XTEA_BLOCK_SIZE 8
#define XTEA_IV_SIZE 8
#define XTEA_KEY_SIZE 16

SemaphoreHandle_t doneEncryptingSemaphore;
SemaphoreHandle_t doneDecryptingSemaphore;

int data_size = 512;
unsigned long encryptFuncStartTime;
unsigned long encryptTaskStartTime;
unsigned long encryptFuncEndTime;
unsigned long encryptTaskEndTime;
unsigned long encryptFuncElapsedMicros;
unsigned long encryptTaskElapsedMicros;
unsigned long encryptElapsedMicros;

unsigned long decryptFuncStartTime;
unsigned long decryptTaskStartTime;
unsigned long decryptFuncEndTime;
unsigned long decryptTaskEndTime;
unsigned long decryptFuncElapsedMicros;
unsigned long decryptTaskElapsedMicros;
unsigned long decryptElapsedMicros;

uint8_t iv[XTEA_IV_SIZE] = {0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07};

uint8_t key[XTEA_KEY_SIZE] = {
    0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
    0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x0F, 0x10};

char packet[512];
char packet16[] = "sixteenbytetext";
char packet32[] = "sixteenbytetextsixteenbytestext";
char packet64[] = "sixteenbytetextsixteenbytestextsixteenbytestextsixteenbytestext";
char packet128[] = "sixteenbytetextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestext";
char packet256[] = "sixteenbytetextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestext";
char packet512[] = "sixteenbytetextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestext";
int packetSize = 0;

uint8_t resultBuffer[512];

void time_output() {
    Serial.print("E ");
    Serial.print(encryptElapsedMicros);
    Serial.print("  D ");
    Serial.println(decryptElapsedMicros);
}

void decrypted_output(uint8_t* buffer, size_t size) {
    Serial.print("Decrypted: ");
    for (size_t i = 0; i < size; ++i) {
        Serial.print((char)buffer[i]);
    }
    Serial.println();
}

void EncryptTask( void *pvParameters ){    
     if (xSemaphoreTake(doneDecryptingSemaphore, portMAX_DELAY) == pdTRUE){  
        encryptFuncStartTime = micros();  
        xtea.ecbEncrypt(key, resultBuffer, data_size);
        xSemaphoreGive(doneEncryptingSemaphore);
        encryptFuncEndTime = micros();
        encryptFuncElapsedMicros = encryptFuncEndTime - encryptFuncStartTime;
        vTaskDelete(NULL);
    }
}

void DecryptTask( void *pvParameters ){
    if (xSemaphoreTake(doneEncryptingSemaphore, portMAX_DELAY) == pdTRUE) {
        decryptFuncStartTime = micros();
        xtea.ecbDecrypt(key, resultBuffer, data_size);
        xSemaphoreGive(doneDecryptingSemaphore);
        decryptFuncEndTime = micros();
        decryptFuncElapsedMicros = decryptFuncEndTime - decryptFuncStartTime;
        vTaskDelete(NULL);
    }
}

void setup() {
    doneEncryptingSemaphore = xSemaphoreCreateBinary();
    doneDecryptingSemaphore = xSemaphoreCreateBinary();
    delay(10000);

    if (data_size == 16) {
        strcpy(packet, packet16);
    } else if (data_size == 32) {
        strcpy(packet, packet32);
    } else if (data_size == 64) {
        strcpy(packet, packet64);
    } else if (data_size == 128) {
        strcpy(packet, packet128);
    } else if (data_size == 256) {
        strcpy(packet, packet256);
    } else if (data_size == 512) {
        strcpy(packet, packet512);
    }

    Serial.begin(9600);
    Serial.println(" ");
    Serial.print("OS || ");
    Serial.print(clock_get_hz(clk_sys) / 1000000);
    Serial.print("mhz || ");
    Serial.print("XTEA");
    packetSize = strlen(packet);
    Serial.print(" || datasize ");
    Serial.println(packetSize + 1);
    memcpy(resultBuffer, packet, data_size);
    xSemaphoreGive(doneDecryptingSemaphore);
}

void loop() {
    encryptTaskStartTime = micros();
    xTaskCreate(
    EncryptTask
    ,  "encrypt"   
    ,  10000
    ,  NULL
    ,  1  
    ,  NULL );
    encryptTaskEndTime = micros();
    delay(1000);
    encryptTaskElapsedMicros = encryptTaskEndTime - encryptTaskStartTime;
    encryptElapsedMicros = encryptTaskElapsedMicros + encryptFuncElapsedMicros;
    Serial.print("E ");
    Serial.print(encryptElapsedMicros);

    decryptTaskStartTime = micros();
    xTaskCreate(
    DecryptTask
    ,  "decrypt"   
    ,  10000
    ,  NULL
    ,  1  
    ,  NULL );
    decryptTaskEndTime = micros();
    delay(1000);
    decryptTaskElapsedMicros = decryptTaskEndTime - decryptTaskStartTime;
    decryptElapsedMicros = decryptTaskElapsedMicros + decryptFuncElapsedMicros;
    Serial.print("  D ");
    Serial.println(decryptElapsedMicros);
    decrypted_output(resultBuffer, data_size);
    delay(3000);
}

File: XTEA/MicroPython/XTEA with OS.py
import pyRTOS
import machine
import utime
import uos
from xtea_cipher import *
from time import sleep
baton = pyRTOS.mutex()
freq = 200000000
printfreq = int(str(freq)[:3])
datasize = 16
plain = ""
plain16 = "sixteenbytetext"
plain32 = "sixteenbytetextsixteenbytestext"
plain64 = "sixteenbytetextsixteenbytestextsixteenbytestextsixteenbytestext"
plain128 = "sixteenbytetextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestext"
plain256 = "sixteenbytetextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestext"
plain512 = "sixteenbytetextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestext"

key = bytearray([
    0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 
    0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x0F, 0x10
])

iv = bytearray([
    0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07
])
encryptedString = ""
encryptFuncStartTime = 0
encryptTaskStartTime = 0
encryptFuncEndTime = 0
encryptTaskEndTime = 0
encryptFuncElapsedMicros = 0
encryptTaskElapsedMicros = 0
encryptElapsedMicros = 0
machine.freq(freq)
print("With OS || ", printfreq, "mhz || ", "XTEA", " || datasize ", datasize)

def encryptData(data):
	global encryptFuncStartTime, encryptFuncEndTime, encryptFuncElapsedMicros
	encryptFuncStartTime = utime.ticks_us()
    data_bytes = bytes(data, 'utf-8')
    cipher = aes(key, MODE_ECB, iv)
    encryptedData = cipher.encrypt(data_bytes)
	encryptFuncEndTime = utime.ticks_us()
	encryptFuncElapsedMicros = encryptFuncEndTime - encryptFuncStartTime
    return encryptedData

def decryptData(encryptedData):
    decipher = xtea(key, MODE_ECB, iv)
    decryptedData = decipher.decrypt(encryptedData)
    return decryptedData.strip().decode('utf-8')

def enc_task(self):
	global encryptedString, encryptTaskStartTime, encryptTaskEndTime, encryptTaskElapsedMicros
    baton.lock()
    encryptTaskStartTime = utime.ticks_us()
    global encryptedString
    encryptedString = encryptData(plain)
    encryptTaskEndTime = utime.ticks_us()
    encryptTaskElapsedMicros = encryptTaskEndTime - encryptTaskStartTime
    baton.unlock()
    return encryptedString
	yield

def dec_task(self):
    baton.lock()
    global encryptedString
    baton.unlock()
    return decryptedString
	yield
        
if datasize == 16:
    plain = plain16
elif datasize == 32:
    plain = plain32
elif datasize == 64:
    plain = plain64
elif datasize == 128:
    plain = plain128
elif datasize == 256:
    plain = plain256
elif datasize == 512:
    plain = plain512		

while True:
	yield [pyRTOS.timeout(0.5)]
		if self.name == "enc_task":
			target = "dec_task"
		else:
			target = "enc_task"
		
	yield [pyRTOS.wait_for_message(self)]
	
	pyRTOS.add_task(pyRTOS.Task(enc_task, name="encryption",mailbox=True))
	pyRTOS.add_task(pyRTOS.Task(dec_task, name="encryption",mailbox=True))
	pyRTOS.start()

    encryptElapsedMicros = encryptTaskElapsedMicros + encryptFuncElapsedMicros
#    print(decryptedString)
    print('E', encryptElapsedMicros)
    utime.sleep_ms(3000)
